File: numix-helper.rb
#!/usr/bin/env ruby

# Dependencies:
# - json
# - color
# - commander
require 'json'
require 'color'
require 'commander/import'


SHAPES = %w(circle square)

def notify_send message
	`notify-send -i numix '#{message}'`
end

program :name, 'Numix Helper'
program :version, '0.0.1'
program :description, 'Helper scripts for Numix icon development'
program :help, 'Author', 'Mészáros Máté Róbert <mateusz.meszaros@gmail.com>'

global_option('-n', '--notify', 'Send OS notification when the command finished')

default_command :help


def puts_color color, fmt
	puts fmt % {rgb: color.to_rgb.hex, hsl: color.css_hsl}
end

def calc_value base, change
	if change =~ /^[+-]/
		base + change.to_f
	else
		change.to_f
	end
end

command :color do |c|
	c.syntax = 'numix color HEX_COLORS...'
	c.option '-H', '--hue=H', String, 'Rotate the hue by the given amount (0-360)'
	c.option '-S', '--saturation=S', String, 'Change the saturation by the given amount (0-100)'
	c.option '-L', '--luminosity=L', String, 'Change the luminosity by the given amount (0-100)'
	c.option '-f', '--format=FORMAT', String, 'Format the output with the given ruby format string'
	c.action do |args, options|
		options.default :format => '| #%{rgb} | %{hsl} |'
		args.each do |color|
			hsl = Color::RGB.by_hex(color).to_hsl
			puts_color hsl, options.format
			hsl.hue = calc_value(hsl.hue, options.hue) if options.hue
			hsl.saturation = calc_value(hsl.saturation, options.saturation) if options.saturation
			hsl.luminosity = calc_value(hsl.luminosity, options.luminosity) if options.luminosity
			puts_color hsl, options.format
		end
	end
end

command :blend do |c|
	c.syntax = 'numix blend COLOR1 COLOR2 [options]'
	c.option '-f', '--format=FORMAT', String, 'Format the output with the given ruby format string'
	c.option '-r', '--ratio=RATIO', Array, 'Use given ratios to blend the colors'
	c.action do |args, options|
		options.default :format => '| %{r} | %{rgb} | %{hsl} |', :ratio => [0.5]
		if args.size != 2
			puts "[ERROR]: The blend command expects exactly 2 arguments!"
		else
			c1 = Color::RGB.by_hex(args[0]).to_hsl
			c2 = Color::RGB.by_hex(args[1]).to_hsl
			options.ratio.each do |r|
				mixture = c1.mix_with(c2, r.to_f)
				puts options.format % {r: r, rgb: mixture.to_rgb.hex, hsl: mixture.css_hsl}
			end
		end
	end
end


command :setup do |c|
	c.syntax = 'numix setup ICON_NAME [options]'
	c.description = 'Set up links for a new icon'
	c.option '-r', '--rename LINK_NAME', String, 'Rename the link (Defaults to ICON_NAME)'
	c.option '-b', '--branch', 'Create git branch named LINK_NAME if not yet created'
	c.action do |args, options|
		icon_name = args.first
		link_name = options.rename || icon_name
		`git checkout -b #{link_name}` if options.branch && `git branch --list #{link_name}`.empty?
		Dir.mkdir('ln') unless File.directory?('ln')
		SHAPES.each do |shape|
			target = "icons/#{shape}/48/#{icon_name}.svg"
			`cp templates/#{shape}/48.svg #{target}` unless File.file? target
			`ln -s ../#{target} ln/#{link_name}.#{shape}.svg`
		end
		notify_send 'Setup finished!' if options.notify
	end
end


DATA_JSON = 'data.json'

class Hash
	def sorted!
		self.keys.sort_by{ |k| k.downcase }.each { |k| self[k] = self.delete k }
	end
end

class Array
	def insert_sorted item
		idx = (0...self.size).bsearch{ |i| item.downcase < self[i].downcase } || self.size
		self.insert idx, item
	end
end

class IconEntry
	def initialize entry
		@hash = entry
	end

	def linux icon
		unless @hash.include? 'linux'
			@hash['linux'] = { 'root' => icon }
			@hash.sorted!
		else
			@hash['linux']['symlinks'] = [] unless @hash['linux'].include? 'symlinks'
			@hash['linux']['symlinks'].insert_sorted icon
		end
	end

	def android icon
		unless @hash.include? 'android'
			@hash['android'] = []
			@hash.sorted!
		end
		@hash['android'].insert_sorted icon
	end
end

class IconData
	def initialize file
		@file = file
		@data = JSON[File.read file]
	end

	def [] icon
		@data[icon] = {} unless @data.include? icon
		IconEntry.new @data[icon]
	end

	def rename old_name, new_name
		@data[new_name] = @data.delete old_name
	end

	def save!
		@data = @data.sort_by{ |k,v| k.downcase }.to_h
		json = JSON.pretty_generate @data, indent: "\t"
		File.open(@file, 'w') { |file| file.write(json + "\n") }
	end
end

command :data do |c|
	c.syntax = 'numix data ICON_NAME [options]'
	c.description = 'Add data entry for the icon'
	c.option '-l', '--linux=ENTRY', String, 'Add linux entry (root or symlink)'
	c.option '-a', '--android=ENTRY', String, 'Add android entry (root or symlink)'
	c.option '-r', '--rename=ICON_NAME', String, 'Rename the data entry base name'
	c.action do |args, options|
		icon_name = args.first
		if options.linux or options.android or options.rename
			data = IconData.new DATA_JSON
			data[icon_name].linux options.linux if options.linux
			data[icon_name].android options.android if options.android
			data.rename icon_name, options.rename if options.rename
			data.save!
		end
	end
end


command :render do |c|
	c.syntax = 'numix render ICON_NAME ...'
	c.description = 'Render svg icons to raster images (png)'
	c.option '-s', '--size SIZES', Array, 'Specify the resolution of the raster.'
	c.option '-b', '--bundle', 'Create bundle images with all the shapes'
	c.option '-q', '--quiet', 'Diable all unnecessary program output'
	c.action do |icon_names, options|
		options.default :size => [48]
		sizes = options.size.map(&:to_i).sort.reverse
		icon_names.each do |icon_name|

			puts "Rendering '#{icon_name}' ..." unless options.quiet
			pngs = sizes.map{ |s| [s, []] }.to_h
			sizes.each do |size|
				SHAPES.each do |shape|
					svg = "icons/#{shape}/48/#{icon_name}.svg"
					png = "#{icon_name}.#{shape}.#{size}.png"
					if File.file? svg
						`inkscape #{svg} -e #{png} -w #{size}`
						puts "... [DONE] #{png}" unless options.quiet
						pngs[size].push png
					else
						puts "... [MISS] #{svg}" unless options.quiet
					end
				end
			end

			if options.bundle
				parts = []
				sizes.each do |size|
					source = pngs[size].join(' ')
					target = "#{icon_name}.#{size}.png"
					padding = (sizes.max - size) / 2.0
					`convert #{source} -bordercolor none -border #{padding}x4 -rotate 90 -append -rotate -90 #{target}`
					puts "... [BUNDLE] #{target}" unless options.quiet
					parts.push target
				end
				if 1 < sizes.size
					source = parts.join(' ')
					target = "#{icon_name}.#{sizes.join('-')}.png"
					`convert #{source} -append #{target}`
					puts "... [BUNDLE] #{target}" unless options.quiet
					`rm #{source}` if options
				end
			end

		end
		notify_send 'Rendering finished!' if options.notify
	end
end

File: numix-helper.py
#!/usr/bin/env python
import click
import json
from colormath.color_objects import sRGBColor, HSLColor
from colormath.color_conversions import convert_color
# colormath
# - seems needs system python - https://github.com/pyenv/pyenv/issues/1474
# - mixing / blending is not supported - https://github.com/gtaylor/python-colormath/issues/4
import subprocess
from os import makedirs, symlink, system
from os.path import exists, expanduser
from shutil import copy
from glob import glob


@click.group()
def numix_helper():
	pass


# TODO: add global --version option
# TODO: move --quiet option to a global level
# TODO: add global -v/--verbose option
# TODO: make render --size and --shape accept comma separated options
# TODO: add autocompletion config
# - https://click.palletsprojects.com/en/7.x/bashcomplete/#activation-script
# TODO: add testing
# - https://click.palletsprojects.com/en/7.x/api/#testing
# TODO: try creating a point transformer (geometric transformations)
# - https://click.palletsprojects.com/en/7.x/commands/#multi-command-pipelines
# - check out the  in the geometry directory
# TODO: implement blend command if colormath supports it


SHAPES = ['circle', 'square']


def find_icon(name):
	prefix = f"{expanduser('~')}/.local/share/icons/hicolor"
	paths = glob(f"{prefix}/*/apps/{name}.png")
	size_map = {
		int(path[len(prefix)+1:].split('x')[0]): path
		for path in paths
	}
	return size_map[max(size_map.keys())]


@numix_helper.command('setup')
@click.argument('icon_name')
# FIXME renaming this option to --git might make more sense, since we are asking to setup git too
@click.option('create_branch', '-b', '--branch', help='Create and checkout git branch if not yet created', is_flag=True)
@click.option('original_name', '-o', '--original', help='Find the original icon and create a link')
def numix_setup(icon_name, create_branch, original_name):
	"""Set up links, files and git for new icon development"""
	if create_branch:
		system(f"git checkout -b {icon_name}")

	if original_name is not None:
		original_icon_path = find_icon(original_name)
		symlink(original_icon_path, f"{icon_name}.original.png")

	makedirs('ln', exist_ok=True)
	for shape in SHAPES:
		target = f"icons/{shape}/48/{icon_name}.svg"
		if not exists(target):
			copy(f"templates/{shape}/48.svg", target)
		link_path = f"ln/{icon_name}.{shape}.svg"
		if not exists(link_path):
			symlink(f"../{target}", link_path)


def git_current_branch_name():
	return subprocess.check_output(["git", "rev-parse", "--abbrev-ref", "HEAD"]).decode('utf-8').strip()


def print_if(predicate, msg):
	if predicate:
		print(msg)


@numix_helper.command('render')
@click.argument('icon_names', nargs=-1)
@click.option('sizes', '-s', '--size', help='Specify the resolution of the raster.', multiple=True, type=int, default=[48], show_default=True)
@click.option('shapes', '-S', '--shape', help='Specify the shapes for which to render.', multiple=True, type=click.Choice(SHAPES), default=SHAPES, show_default=True)
@click.option('bundle', '-B', '--bundle', help='Create bundle images with all the shapes', is_flag=True)
@click.option('quiet', '-q', '--quiet', help='Diable all extraneous program output', is_flag=True)
def numix_render(icon_names, sizes, shapes, bundle, quiet):
	"""Render svg icons to raster images (png)"""
	if bundle:
		raise click.UsageError("Bundling is not implemented yet! Please contact the developer.")
	if len(icon_names) < 1:
		icon_names = [git_current_branch_name()]
	for icon_name in icon_names:
		print_if(not quiet, f"Rendering '{icon_name}' ...")
		for size in sizes:
			for shape in shapes:
				svg = f"icons/{shape}/48/{icon_name}.svg"
				png = f"{icon_name}.{shape}.{size}.png"
				if exists(svg):
					stdout = subprocess.check_output(["inkscape", svg, "-o", png, "-w", size, "2>&1"])
					# TODO log if verbose
					print_if(not quiet, f"... [DONE] {png}")
				else:
					print_if(not quiet, f"... [MISS] {png}")


def add_or_replace(base_value, input_value):
	if input_value.startswith(('+', '-')):
		return base_value + float(input_value)
	return float(input_value)


@numix_helper.command('color')
@click.argument('hex_colors', nargs=-1)
@click.option('hue', '-H', '--hue', help='Rotate the hue by/to the given value (0-360)')
@click.option('saturation', '-S', '--saturation', help='Change the saturation by/to the given value (0-100)')
@click.option('luminosity', '-L', '--luminosity', help='Change the luminosity by/to the given value (0-100)')
def numix_color(hex_colors, hue, saturation, luminosity):
	"""Change and display colors in various color spaces"""
	for hex_color in hex_colors:
		rgb = sRGBColor.new_from_rgb_hex(hex_color)
		hsl = convert_color(rgb, HSLColor)
		print(f'| {rgb.get_rgb_hex()} | hsl({hsl.hsl_h:.2f}, {hsl.hsl_s*100:.2f}%, {hsl.hsl_l*100:.2f}%) |')
		if hue:
			new_h = add_or_replace(hsl.hsl_h, hue)
			# Rotate back to 0-360
			hsl.hsl_h = new_h + 360*(new_h<0) - 360*(new_h>360)
		if saturation:
			new_s = add_or_replace(hsl.hsl_s*100, saturation)
			# Cut and scale into 0-1
			hsl.hsl_s = max(0, min(new_s, 100)) / 100
		if luminosity:
			new_l = add_or_replace(hsl.hsl_l*100, luminosity)
			# Cut and scale into 0-1
			hsl.hsl_l = max(0, min(new_l, 100)) / 100
		rgb = convert_color(hsl, sRGBColor)
		print(f'| {rgb.get_rgb_hex()} | hsl({hsl.hsl_h:.2f}, {hsl.hsl_s*100:.2f}%, {hsl.hsl_l*100:.2f}%) |')


class IconEntry:
	def __init__(self, entry):
		self._entry = entry

	def add_linux(self, name):
		if 'linux' not in self._entry:
			self._entry['linux'] = {'root': name}
		else:
			names = self._entry['linux'].get('symlinks', [])
			names.append(name)
			self._entry['linux']['symlinks'] = sorted(set(names))

	def add_android(self, name):
		names = self._entry.get('android', [])
		names.append(name)
		self._entry['android'] = sorted(set(names))


class IconDataBase:
	def __init__(self, file):
		self._file = file
		with open(file, 'r') as f:
			self._data = json.load(f)

	def __getitem__(self, name):
		if name not in self._data:
			self._data[name] = {}
		return IconEntry(self._data[name])

	def rename(self, name, new_name):
		self._data[new_name] = self._data.pop(name)

	def save(self):
		with open(self._file, 'w') as f:
			json.dump(self._data, f, indent=4, sort_keys=True)


@numix_helper.command('data')
@click.argument('icon_name')
@click.option('linux_entry', '-l', '--linux', help='Add linux (desktop icon) entry.')
@click.option('android_entry', '-a', '--android', help='Add android (app activity) entry.')
@click.option('new_icon_name', '-r', '--rename', help='Rename the icon entry.')
@click.option('registry_file', '-f', '--file', type=click.Path(exists=True), default='data.json', show_default=True, help='Use custom registry file.')
def numix_data(icon_name, linux_entry, android_entry, new_icon_name, registry_file):
	"""Add an entry to the icon name registry (data.json)"""
	if linux_entry or android_entry or new_icon_name:
		db = IconDataBase(registry_file)
		if linux_entry is not None:
			db[icon_name].add_linux(linux_entry)
		if android_entry is not None:
			db[icon_name].add_android(android_entry)
		if new_icon_name is not None:
			db.rename(icon_name, new_icon_name)
		db.save()
	else:
		print(f"Warning: no action specified, no changes made.")


if __name__ == '__main__':
	numix_helper()
